README.md
# library-shipper
Library of processing functions for Magento 2 ShipperHQ client

More information on ShipperHQ capabilities are available at www.shipperhq.com

Facts
-----
- [extension on GitHub](https://github.com/shipperhq/module-shipper)
- Magento v1.0 available for download from www.shipperhq.com

Description
-----------
This library consists of classes used by the ShipperHQ Magento 2 extension to send and processing shipping rate requests.
It is not intended to be installed alone and therefore does not require standalone installation instructions. It is a dependency of shipperhq/module-shipper. 

Compatibility
-------------
- Magento >= 2.0 (Incl 2.2.x)

- Supports both Magento Opensource (Community) and Magento Commerce (Enterprise)

Support
-------
If you have any issues with this library, open an issue on [GitHub](https://github.com/shipperhq/library-shipper/issues).
Alternatively you can contact us via email at support@shipperhq.com or via our website https://shipperhq.com/contact

Contribution
------------
Any contribution is highly appreciated. The best way to contribute code is to open a [pull request on GitHub](https://help.github.com/articles/using-pull-requests).

License
-------
See license files

Copyright
---------
Copyright (c) 2015 Zowta LLC (http://www.ShipperHQ.com)



CHANGELOG.MD
## 20.3.3
SHQ16-2392 added comments and refactored unit tests, added tests for calendarDetails default date, M2-52 remove dependency on debug setting to enable logging


## 20.7.0 (2018-06-21)
Bring releases and versioning back in line


## 20.8.0 (2018-06-25)
SHQ18-277 refactor save selections functions to reuse selection object


## 20.9.0 (2018-08-02)
SHQ18-155 support map and location details for in store pickup


## 20.10.0 (2019-03-14)
SHQ18-1613 Display actual method chosen by rate shopping in admin panel


## 20.10.1 (2019-04-29)
SHQ18-1675 Added support for the new TimeSlotBreakDown field


## 20.11.0 (2019-06-12)
SHQ18-2107 Support for allowed methods for multiple API keys on one store


## 20.11.1 (2019-09-12)
SHQ18-2587/SHQ18-2613 Add price less fees to rates where available


## 20.12.0 (2019-11-20)
SHQ18-2869 Process rate response as array


## 20.12.1 (2019-12-10)
SHQ18-2990 Copy object to array function into library


## 20.12.2 (2020-02-26)
MNB-48 Support for storing rate shopped methods shipments as order comment


## 20.12.3 (2020-06-12)
MNB-422 Revalidate saved addresses that have invalid validation status


## 20.12.4 (2020-07-16)
MNB-484 Ensure default date in calendar uses timezone


## 20.13.0 (2020-08-20)
MNB-604 Add support for box comments for merged rates


## 20.14.0 (2021-03-29)
RIV-443 Add place order endpoint


## 20.14.1 (2021-06-07)
MNB-1251 Fix around handling no rates


## 20.14.2 (2021-07-08)
MNB-1227 Fix around promotional rules


## 20.14.3 (2021-07-30)
 MNB-1492 Ensure city is set in request on date or option change 


## 20.15.0 (2021-11-19)
RIV-687 Add freight quote ID to carrier group


## 20.16.0 (2022-05-13)
MNB-2430 M2.4.4 compatibility


## 20.17.0 (2022-11-29)
MNB-3219 Fix issue with M2 sales order grid showing incorrect year on dates


## 20.17.1 (2022-12-12)
MNB-2930 Fix order notes with generic carrier


## 20.17.2 (2022-12-14)
MNB-3119 fix broken checkout with one day calendar




src/Type/BaseCalendar.php
<?php
/**
 *
 * ShipperHQ Shipping Module
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magentocommerce.com so we can send you a copy immediately.
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade Magento to newer
 * versions in the future. If you wish to customize Magento for your
 * needs please refer to http://www.magentocommerce.com for more information.
 *
 * ShipperHQ Shipping
 *
 * @category ShipperHQ
 * @package ShipperHQ_Lib
 * @copyright Copyright (c) 2014 Zowta LLC (http://www.ShipperHQ.com)
 * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @author ShipperHQ Team sales@shipperhq.com
 */
namespace ShipperHQ\Lib\Type;

use ShipperHQ\Lib\Checkout\AbstractService;
use ShipperHQ\Lib\Helper\Date;
/**
 * Class BaseCalendar
 * @package ShipperHQ_Lib
 */
class BaseCalendar
{
    /**
     * @var AbstractService
     */
    protected $checkoutService;
    /**
     * @var \ShipperHQ\Lib\AdminOrder\AbstractService
     */
    protected $adminOrderService;
    /**
     * AbstractService $calendarService
     * @codeCoverageIgnore
     */
    public function __construct(
        AbstractService $checkoutService,
        \ShipperHQ\Lib\AdminOrder\AbstractService $adminOrderService
    ) {
        $this->checkoutService = $checkoutService;
        $this->adminOrderService = $adminOrderService;
    }

    /*
     * Process date select action
     */
    public function processDateSelect(
        $dateSelected,
        $carrierId,
        $carrierCode,
        $carrierGroupId,
        $addressData,
        $cartId,
        $addressId = false
    ) {
        $params = $this->getDateSelectSaveParameters($dateSelected, $carrierId, $carrierCode, $carrierGroupId);
        $this->checkoutService->saveSelectedData($params);
        $this->checkoutService->cleanDownRates($cartId, $carrierCode, $carrierGroupId, $addressId);
        $addressArray = [
            'street'     => $addressData->getStreet(),
            'city'       => $addressData->getCity(),
            'region'     => $addressData->getRegion(),
            'region_id'  => $addressData->getRegionId(),
            'postcode'   => $addressData->getPostcode(),
            'country_id' => $addressData->getCountryId()
        ];
        try {
            $rates = $this->checkoutService->reqeustShippingRates(
                $cartId,
                $carrierCode,
                $carrierGroupId,
                $addressArray,
                $addressId
            );
        } catch (\Exception $e) {
            // handle so we can clean down rates if necessary
        }
        // need to do smart cleaning at this point
        $this->checkoutService->cleanDownSelectedData();
        return $rates;
    }

    /*
    * Process date select action
    */
    public function processAdminDateSelect(
        $dateSelected,
        $carrierId,
        $carrierCode,
        $carrierGroupId,
        $cartId
    ) {
        $params = $this->getDateSelectSaveParameters($dateSelected, $carrierId, $carrierCode, $carrierGroupId);
        $this->adminOrderService->saveSelectedData($params);
        $this->adminOrderService->cleanDownRates($cartId, $carrierCode, $carrierGroupId);
        $rates = [];
        try {
            $rates = $this->adminOrderService->reqeustShippingRates(
                $cartId,
                $carrierCode,
                $carrierGroupId
            );
        } catch (\Exception $e) {
            // handle so we can clean down rates if necessary
        }
        // need to do smart cleaning at this point
        $this->adminOrderService->cleanDownSelectedData();
        return $rates;
    }

    public function saveDateSelectOnCheckoutProceed($dateSelected, $carrierId, $carrierCode, $carrierGroupId)
    {
        $params = $this->getDateSelectSaveParameters($dateSelected, $carrierId, $carrierCode, $carrierGroupId);
        $this->checkoutService->saveSelectedData($params);
    }

    public function resetDateSelectOnCheckoutProceed()
    {
        $this->checkoutService->cleanDownSelectedData(['SelectedDate']);
    }

    public function getDateSelectSaveParameters($dateSelected, $carrierId, $carrierCode, $carrierGroupId)
    {
        $selections = [
            'CarrierGroupId' => $carrierGroupId,
            'CarrierId' => $carrierId,
            'CarrierCode' => $carrierCode,
            'SelectedDate' => $dateSelected
        ];
        return $selections;
    }

    public function processCalendarDetails($carrierRate, $carrierGroupDetail)
    {
        $calendarDetails = $carrierRate['calendarDetails'];
        // need to convert this to cldrFormat
        $locale = $carrierGroupDetail['locale'] ?? null;
        $deliveryDateFormat = $carrierRate['deliveryDateFormat'];

        foreach ($carrierRate['rates'] as $rate) {
            $defaultDate = intval($rate['deliveryDate']/1000);
            break;
        }
        if (!empty($calendarDetails)) {
            $calendarDetails = $this->getCalendarDetailsArray(
                $calendarDetails,
                $carrierGroupDetail,
                $carrierRate['carrierId'],
                $carrierRate['carrierCode'],
                $locale,
                $deliveryDateFormat,
                $defaultDate
            );
        }

        return $calendarDetails;
    }

    public function getCalendarDetailsArray(
        $calendarDetails,
        $carrierGroupDetail,
        $carrierId,
        $carrierCode,
        $locale,
        $deliveryDateFormat,
        $defaultDate
    ) {
        $calendarDetails['dateFormat'] = Date::getDateFormat($locale);
        $calendarDetails['datepickerFormat'] = Date::getDatepickerFormat($locale);
        $calendarDetails['displayDateFormat'] = Date::getCldrDateFormat($locale, $deliveryDateFormat);
        $calendarDetails['timezone'] = $carrierGroupDetail['timezone'];
        $calendarDetails['default_date_timestamp'] = $defaultDate;
        // SHQ16-2041 pass default date in calendar details
        $calendarDetails['default_date'] = $this->getDateFromTimestamp(
            $defaultDate,
            $calendarDetails['timezone'], // MNB-484, TB Reversed a change from SHQ16-2078
            $calendarDetails['dateFormat']
        );
        if ($calendarDetails['startDate'] != '') {
            $calendarDetails['start'] = intval($calendarDetails['startDate']/1000);
        } else {
            $calendarDetails['start'] = $defaultDate;
        }
        $calendarDetails['carrier_id'] = $carrierId;
        $calendarDetails['carrier_code'] = $carrierCode;
        $dateOptions = $this->getDateOptions($calendarDetails);
        if (count($dateOptions)>0) {
            $deliveryDatesAndTimes = [];
            if (isset($calendarDetails['timeSlots']) && !empty($calendarDetails['timeSlots'])) {
                if (isset($calendarDetails['timeSlotBreakdown']) && !empty($calendarDetails['timeSlotBreakdown'])) {
                    $deliveryDatesAndTimes[$calendarDetails['default_date']] = $this->timeSlotBreakdownListToSlotList($calendarDetails['timeSlotBreakdown']);
                } else {
                    foreach ($dateOptions as $dateKey => $date) {
                        if ($slotsFound = $this->getDeliveryTimeSlots($calendarDetails, $dateKey)) {
                            $deliveryDatesAndTimes[$dateKey] = $slotsFound;
                        } else {
                            unset($dateOptions[$dateKey]);
                        }
                    }
                }
            }

            if (count($dateOptions) <= 0) {
                // handle if no date options found
                $dateOptions = [];
            }

            if (count($deliveryDatesAndTimes) > 0) {
                $calendarDetails['display_time_slots']= $deliveryDatesAndTimes;
                $calendarDetails['showTimeslots'] = true;
            } else {
                $calendarDetails['display_time_slots'] = false;
                $calendarDetails['showTimeslots'] = false;
            }
        } else {
            // MNB-3119 Something has gone wrong
            $calendarDetails['display_time_slots']= false;
            $calendarDetails['showTimeslots'] = false;
            $calendarDetails['min_date'] = $calendarDetails['default_date'];
            $calendarDetails['max_date'] = $calendarDetails['default_date'];
        }
        $calendarDetails['allowed_dates'] = $dateOptions;
        if (!empty($dateOptions)) { // SHQ16-2041
            $keys = array_keys($dateOptions);
            $calendarDetails['min_date'] = $keys[0];
            $calendarDetails['max_date'] = end($keys);
            /*
             * SHQ16-2392 check the selected date is in the date options - if not then remove it and push the default
             * date to the first available. This is an issue for example if you change carriers selected on checkout,
             * the previous carriers selected date may not be viable for this carrier
             */
            $isInDateOptions = in_array($calendarDetails['default_date'], $keys);
            if (!$isInDateOptions) {
                $calendarDetails['default_date'] = $calendarDetails['min_date'];
            }
        }

        return $calendarDetails;
    }

    /**
     * These dates are not influenced by the location
     * @return array an array of dates for Store Pickup
     */
    public function getDateOptions($calendarDetails)
    {
        $numPickupDays = array_key_exists('maxDays', $calendarDetails) ? $calendarDetails['maxDays'] : 10;

        //Convert java linux timestamps (milliseconds) into php linux timestamps (seconds)
        $dateOptions = [];
        $dateFormat = $calendarDetails['dateFormat'];
        $timezone = $calendarDetails['timezone'];
        $startDate = $calendarDetails['start'];
        $endDate = isset($calendarDetails['calendarEndDate']) ?
            intval($calendarDetails['calendarEndDate']/1000)
            : false;

        $arrBlackoutDates = [];
        foreach ($calendarDetails['blackoutDates'] as $blackoutDate) {
            $arrBlackoutDates[] = $this->getDateFromDate($blackoutDate, $timezone, $dateFormat);
        }
        $arrBlackoutDays = $calendarDetails['blackoutDays'];

        if (count($arrBlackoutDays) == 7) {
            // somehow flag that there are no date options available
            return $dateOptions;
        }
        $countDays = 0;
        while ($countDays < $numPickupDays) {
            // support end date inclusive
            if ($endDate && $startDate > $endDate) {
                break;
            }

            $nextDay = $this->getDateFromTimestamp($startDate, $timezone, $dateFormat);

            // Blackout day or date...get next available
            if (in_array($nextDay, $arrBlackoutDates) ||
                in_array($this->getDayOfWeekFromTimestamp($startDate, $timezone), $arrBlackoutDays)) {
                $this->_addDay($startDate);
                $countDays++;
                continue;
            }
            $dateOptions[$nextDay] = $nextDay;
            $this->_addDay($startDate);
            $countDays++;
        }
        return $dateOptions;
    }

    public function timeSlotBreakdownListToSlotList($timeSlotBreakdownList)
    {
        $timeSlots = [];
        foreach ($timeSlotBreakdownList as $timeSlotBreakdown) {
            $key = $this->timeSlotBreakdownToSlotKey($timeSlotBreakdown);
            $value = $this->timeSlotBreakdownToSlotValue($timeSlotBreakdown);
            $timeSlots[$key] = $value;
        }
        return $timeSlots;
    }

    public function timeSlotBreakdownToSlotKey($timeSlotBreakdown)
    {
        return "{$timeSlotBreakdown['timeStart']}_{$timeSlotBreakdown['timeEnd']}";
    }

    public function timeSlotBreakdownToSlotValue($timeSlotBreakdown)
    {
        return "{$timeSlotBreakdown['timeStart']} - {$timeSlotBreakdown['timeEnd']}";
    }

    public function getDeliveryTimeSlots($calendarDetails, $date)
    {
        if (!isset($calendarDetails['timeSlots'])) {
            return false;
        }
        $timezone = $calendarDetails['timezone'];
        $timezoneObj = new \DateTimeZone($timezone);
        $dateFormat = $calendarDetails['dateFormat'];
        $today = $this->getCurrentDate($timezone, $dateFormat);

        $isToday = false;
        $selectedDate = $calendarDetails['default_date'];

        if ($today == $date) {
            $isToday = true;
            // account for same day delivery with lead time in hours
            $exactStartTimeStamp = $calendarDetails['start'];
            // if we are calcuating time slots for the selected date and that date is today.
            // need to now check that the current time is not AFTER the earliest possible start time (including any lead time in hours)
            if ($selectedDate == $today && $calendarDetails['default_date_timestamp'] > $exactStartTimeStamp) {
                $exactStartTimeStamp = $calendarDetails['default_date_timestamp'];
            }
        }

        $currentTime = 0;
        // if we are generating slots for today, make sure we don't offer any in the past,
        // and we account for lead time in hours
        if ($isToday) {
            $currentTimeClass = new \DateTime("now", $timezoneObj);
            $currentTime = $currentTimeClass->getTimestamp();
            $currentTime = max($currentTime, $exactStartTimeStamp);
        }
        // if we are generating slots for the date that matches the default date, there may be lead time in hours,
        // lets account for that by setting the currentTime to be the default_date_timestamp
        if (!$isToday && $selectedDate == $date) {
            $currentTime = $calendarDetails['default_date_timestamp'];
        }

        $timeSlotDetail = (array)$calendarDetails['timeSlots'];

        $sortTime = [];
        foreach ($timeSlotDetail as $key => $val) {
            $values = (array)$val;
            $sortTime[$key] = $values['timeStart'];
            $timeSlotDetail[$key] = $values;
        }

        array_multisort($sortTime, SORT_ASC, $timeSlotDetail);

        // for implementation of date/day based slot detail in future
        $timeSlots = [];
        foreach ($timeSlotDetail as $slotDetail) {
            // If we don't have all our datapoints then drop the slot
            if (!isset($slotDetail['timeStart'], $slotDetail['timeEnd'], $slotDetail['interval'])) {
                continue;
            }

            // Setup slot boundaries and interval
            $slotStartTime = new \DateTime($date . ' ' . $slotDetail['timeStart'], $timezoneObj);
            $slotEndTime = new \DateTime($date . ' ' . $slotDetail['timeEnd'], $timezoneObj);
            $interval = $slotDetail['interval'];

            // if interval is half or full day then calculate those intervals
            if ($interval <= 2) {
                $dayLengthInMins = ($slotEndTime->getTimestamp() - $slotStartTime->getTimestamp()) / 60;
                $interval = $dayLengthInMins / $interval;
            }

            // If end time < start time (slot runs overnight) then clamp it to midnight
            if ($slotEndTime <= $slotStartTime) {
                $slotEndTime->setTime(24, 00);
            }

            for ($e = clone ($s = clone $slotStartTime); $s < $slotEndTime; $s->setTimestamp($e->getTimestamp())) {
                // EndTime = StartTime + Interval
                $e->setTimestamp($s->getTimestamp())
                    ->modify("+ $interval minutes");
                // Don't let end of slot overrun our hard ceiling
                if ($e > $slotEndTime) {
                    $e->setTimestamp($slotEndTime->getTimestamp());
                }

                // If the timeslot has already passed, then move on
                if ($s->getTimestamp() < $currentTime) {
                    continue;
                }

                $intervalStartStr = $s->format('H:i');
                $intervalEndStr = $e->format('H:i');
                $intervalEndStr = ($intervalEndStr === "00:00") ? "24:00" : $intervalEndStr;
                $key = "{$intervalStartStr}_{$intervalEndStr}";
                $value = "{$intervalStartStr} - {$intervalEndStr}";

                $timeSlots[$key] = $value;
            }
            unset($e, $s); // Paranoia
        }

        if (count($timeSlots) == 0) {
            return false;
        }
        ksort($timeSlots);
        return $timeSlots;
    }

    public function getDateFromDate($date, $timezone, $dateFormat)
    {
        // SHQ16-2417 correct black out dates to ignore time zone. This is just to reformat the dates correctly
        $returnDate = $this->getDateFromTimestamp(strtotime($date), 'Europe/London', $dateFormat);
        return $returnDate;
    }

    public function getCurrentDate($timezone, $dateFormat)
    {
        $dateTime = new \DateTime("now", new \DateTimeZone($timezone));
        $returnDate = $dateTime->format($dateFormat);

        return $returnDate;
    }

    public function getDateFromTimestamp($timeStamp, $timezone, $dateFormat)
    {
        $dateTime = new \DateTime("now", new \DateTimeZone($timezone));
        $dateTime->setTimestamp($timeStamp);
        $returnDate = $dateTime->format($dateFormat);

        return $returnDate;
    }

    /**
     * Get a numerical representation of the day of the week from a date
     *
     * @param string $date
     * @return bool|string
     */
    public function getDayOfWeekFromDate($date, $timezone)
    {
        $unixTime = strtotime((string) $date);
        $dayOfWeek = $this->getDateFromTimestamp($unixTime, $timezone, 'N');
        return $dayOfWeek;
    }

    /**
     * Get a numerical representation of the day of the week from a timestamp
     *
     * @param string $date
     * @return bool|string
     */
    public function getDayOfWeekFromTimestamp($timestamp, $timezone)
    {
        $dayOfWeek = $this->getDateFromTimestamp($timestamp, $timezone, 'N');
        return $dayOfWeek;
    }

    /**
     * Given a date will add a day to it.
     *
     * @param     $day
     * @param int $numDaysToAdd
     *
     * @return false|int
     */
    protected function _addDay(&$day, $numDaysToAdd = 1)
    {
        $day = strtotime('+' . $numDaysToAdd . ' day', $day);
        return $day;
    }

    public function getBlackoutDaysList($blackoutArray)
    {
        $arrBlackoutDays = [];
        foreach ($blackoutArray as $dayOfWeek) {
            //Java Sunday = 7, Monday = 1. PHP Monday = 1, Saturday = 6, Sunday = 0
            if ($dayOfWeek == 7) {
                $dayOfWeek = 0;
            }
            $arrBlackoutDays[] = $dayOfWeek;
        }
        return $arrBlackoutDays;
    }
}
